langchain4j-core/src/main/java/dev/langchain4j/web/search/WebSearchEngine.java

```diff
@@ -0,0 +1,45 @@
+package dev.langchain4j.web.search;
+
+import java.util.stream.Collectors;
+
+import static dev.langchain4j.internal.ValidationUtils.ensureNotNull;
+
+/**
+ * Represents a web search engine that can be used to perform searches on the Web in response to a user query.
+ */
+public interface WebSearchEngine {
+
+    /**
+     * Runs a search query on the web search engine and returns a pretty-string representation of the search results.
+     *
+     * @param query the search query
+     * @return a pretty-string representation of the search results
+     */
+    default String runSearch(String query) {
```

Comment:
is this method required? How/where will it be used?

Reply:
This method is used by objects with`@Tool` due `DefaultToolExecutor` is unable to render a custom POJO as response so far.

Reply:
I would avoid having this logic here, it is not a responsibility of `WebSearchEngine` to format results.
If this is needed by a `WebSearchTool`, let's better move it there.

Reply:
makes sense, in fact I had thought about it that way from the beginning and that the `@WebSearchTool` is an abstract class that abstracts, formats and does everything necessary so that any web search that extends this class works automatically as an `@Tool` as well, however `@AiService` does not recognize `@Tools` from an abstract class.

My idea was to do something like this (sorry by the pseudocode):

```
public abstract class WebSearchTool {
  protected abstract WebSearchEngine webSearchEngine();

  @Tool("...")
  public String runSearch(String searchTerm){
    WebSearchResults results = webSearchEngie().search(WebSearchRequest.from(searchTerm));
    return results.prettyPrinting();
  }
}
```
Any implementation of a new web search could be a tool doing something as simple as this **E.g. Google Search Implementation**

```
public class GoogleWebSearchTool extends WebSearchTool {

   private final WebSearchEngine googleSearchEngine;

  //ctr
  GoogleWebSearchTool (....){
     this.googleSearchEngine = GoogleSearchEngine.builder()...
  }
  
  @Override
  protected WebSearchEngine webSearchEngie() {
  return googleSearchEngine;
  }
```

I will change my approach to an interface and test if it works as the first option with the current implementation of `@AiServices`, otherwise I will have to modify the `@AiServices` so that it accepts `@tool` from the parent methods of the tool objects. Any suggestion another suggestion?

Reply:
Hmm... Why not to just make it like this?

```java
public class WebSearchTool {

  private final WebSearchEngine webSearchEngine;

  public WebSearchTool(WebSearchEngine webSearchEngine) {
      this.webSearchEngine = ensureNotNull(webSearchEngine, "webSearchEngine");
  }

  @Tool("...")
  public String runSearch(String searchTerm){
      WebSearchResults results = webSearchEngie().search(WebSearchRequest.from(searchTerm));
      return results.prettyPrinting();
  }
}
```